instant-distance-py/src/lib.rs

```diff
@@ -24,41 +26,84 @@ fn instant_distance_py(_: Python, m: &PyModule) -> PyResult<()> {
     m.add_class::<Search>()?;
     m.add_class::<Hnsw>()?;
     m.add_class::<HnswMap>()?;
+    m.add_class::<DistanceMetric>()?;
     Ok(())
 }
 
+#[pyclass]
+#[derive(Copy, Clone, Default)]
+enum DistanceMetric {
+    #[default]
+    Euclid,
+    Cosine,
+    DotProduct,
+}
+
 #[pyclass]
 struct HnswMap {
-    inner: instant_distance::HnswMap<FloatArray, MapValue>,
+    inner: HnswMapWithMetric,
+}
+
+#[derive(Deserialize, Serialize)]
+enum HnswMapWithMetric {
+    Euclid(instant_distance::HnswMap<FloatArray<EuclidMetric>, MapValue>),
+    Cosine(instant_distance::HnswMap<FloatArray<CosineMetric>, MapValue>),
+    DotProduct(instant_distance::HnswMap<FloatArray<DotProductMetric>, MapValue>),
+}
+
+// Helper macro for dispatching to inner implementation
+macro_rules! impl_for_each_hnsw_with_metric {
+    ($type:ident, $instance:expr, $inner:ident, $($tokens:tt)+) => {
+        match $instance {
+            $type::Euclid($inner) => {
+                $($tokens)+
+            }
+            $type::Cosine($inner) => {
+                $($tokens)+
+            }
+            $type::DotProduct($inner) => {
+                $($tokens)+
+            }
+        }
+    };
 }
 
 #[pymethods]
 impl HnswMap {
     /// Build the index
     #[staticmethod]
     fn build(points: &PyList, values: &PyList, config: &Config) -> PyResult<Self> {
-        let points = points
-            .into_iter()
-            .map(FloatArray::try_from)
-            .collect::<Result<Vec<_>, PyErr>>()?;
-
         let values = values
             .into_iter()
             .map(MapValue::try_from)
             .collect::<Result<Vec<_>, PyErr>>()?;
+        let builder = instant_distance::Builder::from(config);
 
-        let hsnw_map = instant_distance::Builder::from(config).build(points, values);
-        Ok(Self { inner: hsnw_map })
+        let inner = match config.distance_metric {
+            DistanceMetric::Euclid => {
+                let points = FloatArray::try_from_pylist(points)?;
```

Comment:
It looks like we can deduplicate the `FloatArray::try_from_pylist()` and `builder.build()` calls?

Reply:
The underlying types are actually different. They are inferred so not directly visible in the source code but that `FloatArray::try_from_pylist` is actually `FloatArray::<EuclidMetric>::try_from_pylist`. It's a similar story with `builder.build()` - the return type is different in each call.

Reply:
Fair enough!